### PR TITLE
Top 3 Students from each grade

### LINKED ISSUE #6: Top 3 students from each school grade.
For each grade, the 3 best boys and girls get a special announcement. To make this easy the system has to find those.

### PR DESCRIPTION
**Pull request by ghost**
There is still a  //TODO in the method topStudentsMapping. Im working on a solution to but maybe you know to solve the problem. 

Closes #6

### CODE REVIEW

File: api/src/main/java/com/bjs/bjsapi/database/repository/ClassRepository.java [hunk 1]
```diff
@@ -27,6 +26,9 @@
 	@PostFilter("hasRole('ROLE_ADMIN') or @classPermissionEvaluator.hasPermission(authentication,filterObject,'read')")
 	List<Class> findByClassTeacherName(String classTeacherName);
 
+	@PostAuthorize("hasRole('ROLE_ADMIN') or @classPermissionEvaluator.hasPermission(authentication,returnObject.orElse(null),'read')")
```
Comment (aykborstelmann) [inline api/src/main/java/com/bjs/bjsapi/database/repository/ClassRepository.java]:
```suggestion
	@PostFilter("hasRole('ROLE_ADMIN') or @classPermissionEvaluator.hasPermission(authentication,filterObject,'read')")
```

File: api/src/test/java/com/bjs/bjsapi/controllers/TopStudentsControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+    @MockBean
+    private TopStudentsController topStudentsController;
+
+    @BeforeEach
+    void setUp() {
+        topStudentsController = new TopStudentsController(studentRestController, studentRepository, classRepository);
+    }
+
+    @Test
```
Comment (aykborstelmann) [inline api/src/test/java/com/bjs/bjsapi/controllers/TopStudentsControllerTest.java]: There may be a prettier way than making quite similar tests. I will look at this later more

**Review by aykborstelmann — CHANGES_REQUESTED** [review]
This is my first review, I will look at this further later.

**Comment by skyliam** [comment]
Maybe we could return a List filled with the two lists and a method splitting them up?

**Comment by aykborstelmann** [comment]
> Maybe we could return a List filled with the two lists and a method splitting them up?

Yes, I've sort of implemented this way. It will look like 
```json
{
    "male":[
        { }
    ],
    "female":[
        { }
    ]
}
```

File: api/src/main/java/com/bjs/bjsapi/database/model/Class.java [hunk 3]
```diff
@@ -15,7 +15,10 @@
 	@GeneratedValue(strategy = GenerationType.AUTO)
 	private Long id;
 
-	@Column(nullable = false, unique = true)
+	@Column(nullable = false)
+	private String grade;
```
Comment (aykborstelmann) [inline api/src/main/java/com/bjs/bjsapi/database/model/Class.java]: I'm thinking about if this would be better as an integer ... 🤔

**Review by aykborstelmann — CHANGES_REQUESTED** [review]
I think this looks quite good right now. But I will maybe add a few more tests myself. Well done 👍

**Comment by skyliam** [comment]
I look threw your commit later. Christmas is stressing me lately.

**Comment by skyliam** [comment]
I like the names of the Test students 😆

**Comment by aykborstelmann** [comment]
Yeah absolutely take your time. I just wanted to finish my part.

**Review by aykborstelmann — APPROVED** [review]
Well done, I think we can rebase this now

**Pull request merged**